sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/crypto/CryptoAesAndEtmHandler.java

```diff
@@ -0,0 +1,165 @@
+        ByteBuffer byteBuffer = ByteBuffer.wrap(data);
+
+        // Check iv data.
+        int ivLength = byteBuffer.get();
+        if (ivLength != 16) {
+            throw new IllegalArgumentException("Invalid IV length.");
+        }
+        byte[] iv = new byte[ivLength];
+        byteBuffer.get(iv);
+
+        // Check mac data.
+        int macLength = (byteBuffer.get());
+        if (macLength != 32) {
+            throw new IllegalArgumentException("Invalid MAC length.");
+        }
+        byte[] actualHMac = new byte[macLength];
+        byteBuffer.get(actualHMac);
+
+        // Get cipher data.
+        byte[] cipherText = new byte[byteBuffer.remaining()];
+        byteBuffer.get(cipherText);
+
+        // Get secure key and subkeys.
+        byte[] encryptionSubkey = getSubkey(((KeyStore.SecretKeyEntry) keyStoreEntry).getSecretKey(), ENCRYPTION_KEY_LENGTH);
+        byte[] authenticationSubkey = getSubkey(((KeyStore.SecretKeyEntry) keyStoreEntry).getSecretKey(), AUTHENTICATION_KEY_LENGTH);
+
+        // Calculate mac.
+        byte[] expectedHMac = getMacBytes(authenticationSubkey, iv, cipherText);
+
+        // Verity mac value in the encrypted message.
+        if (!MessageDigest.isEqual(expectedHMac, actualHMac)) {
+            throw new SecurityException("Could not authenticate MAC value.");
+        }
+
+        // Decrypt message.
+        CryptoUtils.ICipher cipher = cryptoFactory.getCipher(CryptoConstants.CIPHER_AES, null);
+        cipher.init(DECRYPT_MODE, new SecretKeySpec(encryptionSubkey, KeyProperties.KEY_ALGORITHM_AES), new IvParameterSpec(iv));
+        return cipher.doFinal(cipherText);
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private byte[] getMacBytes(byte[] authKey, byte[] iv, byte[] cipherText) throws InvalidKeyException, NoSuchAlgorithmException {
+        SecretKey macSecureKey = new SecretKeySpec(authKey, KeyProperties.KEY_ALGORITHM_HMAC_SHA256);
+        Mac hMac = Mac.getInstance(KeyProperties.KEY_ALGORITHM_HMAC_SHA256);
+        hMac.init(macSecureKey);
+        hMac.update(iv);
+        hMac.update(cipherText);
+        return hMac.doFinal();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    @VisibleForTesting
+    byte[] getSubkey(@NonNull SecretKey secretKey, int outputDataLength) throws NoSuchAlgorithmException, InvalidKeyException {
+        if (outputDataLength < 1) {
+            throw new IllegalArgumentException("Output data length must be at more than zero.");
+        }
+
+        // Init mac.
+        Mac hMac = Mac.getInstance(KeyProperties.KEY_ALGORITHM_HMAC_SHA256);
+        hMac.init(secretKey);
+
+        // Prepared array and calculate count of iterations.
+        int iterations = (int) Math.ceil(((double) outputDataLength) / ((double) hMac.getMacLength()));
+        if (iterations > 255) {
+            throw new IllegalArgumentException("Output data length must be maximal 255 * hash-length.");
+        }
+
+        // Calculate subkey.
```

Comment:
Is it HKDF algorithm? Shouldn't we specify that here?
Or maybe specify that in the method name?